src/us/kbase/workspace/database/provenance/CreditMetadata.java

```diff
@@ -0,0 +1,346 @@
+				this.resourceType = DEFAULT_RESOURCE_TYPE;
+			}
+
+			try {
+				this.identifier = Common.checkPid(identifier, "identifier", false);
+			} catch (IllegalArgumentException e) {
+				this.errorList.add(e.getMessage());
+			}
+
+			// must be at least one title
+			this.titles = Common.dedupeSimpleList(titles);
+			if (this.titles == null || this.titles.isEmpty()) {
+				this.errorList.add("at least one title must be provided");
+			}
+		}
+
+		/**
+		 * Sets comments for the resource
+		 * @param comments comments as a string
+		 * @return this builder
+		 */
+		public Builder withComments(final String comments) {
+			this.comments = Common.processMultilineString(comments);
+			return this;
+		}
+
+		/**
+		 * Sets the license for the resource
+		 * @param license license as a string
+		 * @return this builder
+		 */
+		public Builder withLicense(final String license) {
+			this.license = Common.processString(license);
+			return this;
+		}
+
+		/**
+		 * Sets the version
+		 * @param withVersion version string
+		 * @return this builder
+		 */
+		public Builder withVersion(final String version) {
+			this.version = Common.processString(version);
+			return this;
+		}
+
+		/**
+		 * Sets the contributor(s)
+		 * @param withContributors list of {@link Contributor}s.
+		 * @return this builder
+		 */
+		public Builder withContributors(final List<Contributor> contributors) {
+			this.contributors = Common.dedupeSimpleList(contributors);
+			return this;
+		}
+
+		/**
+		 * Sets the dates of various resource life cycle events
+		 * @param withDates list of {@link EventDate}s.
+		 * @return this builder
+		 */
+		public Builder withDates(final List<EventDate> dates) {
+			this.dates = Common.dedupeSimpleList(dates);
+			return this;
+		}
+
+		/**
+		 * Sets the funding source(s)
+		 * @param withFunding list of {@link FundingReference}s.
+		 * @return this builder
+		 */
+		public Builder withFunding(final List<FundingReference> funding) {
+			this.funding = Common.dedupeSimpleList(funding);
+			return this;
+		}
+
+		/**
+		 * Sets the related identifiers
+		 * @param withRelatedIdentifiers list of {@link PermanentID}s.
+		 * @return this builder
+		 */
+		public Builder withRelatedIdentifiers(final List<PermanentID> relatedIdentifiers) {
+			this.relatedIdentifiers = Common.dedupeSimpleList(relatedIdentifiers);
+			return this;
+		}
+
+		/**
+		 * Build the {@link CreditMetadata}.
+		 *
+		 * @return the credit metadata.
+		 */
+		public CreditMetadata build() {
+
+			if (version == null && (dates == null || dates.isEmpty())) {
+				errorList.add("must provide either 'version' or one or more 'dates', ideally indicating when the resource was published or when it was last updated");
+			}
+
+			// If the license looks like a URL, ensure it's properly formatted and seems valid.
+			// For these purposes, if it starts with http, assume it's an URL.
+			if (license != null && license.toLowerCase().startsWith("http")) {
```

Comment:
This is a fairly basic (**cough**crap**cough**) way to detect an URL-like string, but should be OK as long as there are instructions to put in the fully-specified URL and examples of such data.

Reply:
moved comment to appropriate line in most recent commit